metriq_gym/benchmarks/mirror_circuits.py

```diff
@@ -0,0 +1,273 @@
+@dataclass
+class MirrorCircuitsData(BenchmarkData):
+    """Data class to store mirror circuits benchmark metadata."""
+    nlayers: int
+    two_qubit_gate_prob: float
+    two_qubit_gate_name: str
+    shots: int
+    num_qubits: int
+    seed: int | None
+    expected_bitstring: str
+
+
+def random_paulis(connectivity_graph: nx.Graph, random_state: random.RandomState) -> QuantumCircuit:
+    """Returns a circuit with randomly selected Pauli gates on each qubit.
+
+    Args:
+        connectivity_graph: Connectivity graph of device to run circuit on.
+        random_state: Random state to select Paulis I, X, Y, Z uniformly at random.
+    """
+    num_qubits = len(connectivity_graph.nodes)
+    qc = QuantumCircuit(num_qubits)
+    
+    paulis = [IGate(), XGate(), YGate(), ZGate()]
+    
+    for qubit in connectivity_graph.nodes:
+        pauli_gate = paulis[random_state.randint(len(paulis))]
+        qc.append(pauli_gate, [qubit])
+    
+    return qc
+
+
+def edge_grab(
+    two_qubit_gate_prob: float,
+    connectivity_graph: nx.Graph,
+    random_state: random.RandomState,
+) -> nx.Graph:
+    """Selects edges for two-qubit gates based on probability and connectivity.
+    
+    Args:
+        two_qubit_gate_prob: Probability of an edge being chosen.
+        connectivity_graph: The connectivity graph for the backend.
+        random_state: Random state to select edges uniformly at random.
+
+    Returns:
+        A graph with selected edges for two-qubit gates.
+    """
+    connectivity_graph = connectivity_graph.copy()
+    candidate_edges = nx.Graph()
+    final_edges = nx.Graph()
+    final_edges.add_nodes_from(connectivity_graph.nodes)
+
+    while connectivity_graph.edges:
+        edges = list(connectivity_graph.edges)
+        if not edges:
+            break
+        num = random_state.randint(len(edges))
+        curr_edge = edges[num]
+        candidate_edges.add_edge(*curr_edge)
+        connectivity_graph.remove_nodes_from(curr_edge)
+
+    for edge in candidate_edges.edges:
+        if random_state.uniform(0.0, 1.0) < two_qubit_gate_prob:
+            final_edges.add_edge(*edge)
+    
+    return final_edges
+
+
+def random_cliffords(
+    connectivity_graph: nx.Graph,
+    random_state: random.RandomState,
+    two_qubit_gate_name: str = "CNOT",
+) -> QuantumCircuit:
+    """Generate random Clifford gates on selected edges.
+    
+    Args:
+        connectivity_graph: Graph with edges for two-qubit gates.
+        random_state: Random state to choose Cliffords uniformly at random.
+        two_qubit_gate_name: Two-qubit gate to use ("CNOT" or "CZ").
+
+    Returns:
+        A circuit with two-qubit gates and random single-qubit Cliffords.
+    """
+    if not connectivity_graph.nodes:
+        return QuantumCircuit(0)
+    
+    num_qubits = max(connectivity_graph.nodes) + 1
+    qc = QuantumCircuit(num_qubits)
+    
+    # Add two-qubit gates
+    two_qubit_gate = CXGate() if two_qubit_gate_name == "CNOT" else CZGate()
```

Comment:
Do we want to either enum these choices or do a check? It might be misleading if someone puts in another string value expecting another gate but then gets either a CX or CZ instead.